src/iceberg/catalog/rest/catalog.h

```diff
@@ -0,0 +1,110 @@
+
+/// \file iceberg/catalog/rest/catalog.h
+/// RestCatalog implementation for Iceberg REST API.
+
+namespace iceberg::rest {
+
+class ICEBERG_REST_EXPORT RestCatalog : public Catalog {
+ public:
+  RestCatalog(const RestCatalog&) = delete;
+  RestCatalog& operator=(const RestCatalog&) = delete;
+  RestCatalog(RestCatalog&&) = delete;
+  RestCatalog& operator=(RestCatalog&&) = delete;
+
+  /// \brief Create a RestCatalog instance
+  ///
+  /// \param config the configuration for the RestCatalog
+  /// \return a unique_ptr to RestCatalog instance
+  static Result<std::unique_ptr<RestCatalog>> Make(const RestCatalogConfig& config);
+
+  std::string_view name() const override;
+
+  Result<std::vector<Namespace>> ListNamespaces(const Namespace& ns) const override;
+
+  Status CreateNamespace(
+      const Namespace& ns,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  Result<std::unordered_map<std::string, std::string>> GetNamespaceProperties(
+      const Namespace& ns) const override;
+
+  Status DropNamespace(const Namespace& ns) override;
+
+  Result<bool> NamespaceExists(const Namespace& ns) const override;
+
+  Status UpdateNamespaceProperties(
+      const Namespace& ns, const std::unordered_map<std::string, std::string>& updates,
+      const std::unordered_set<std::string>& removals) override;
+
+  Result<std::vector<TableIdentifier>> ListTables(const Namespace& ns) const override;
+
+  Result<std::unique_ptr<Table>> CreateTable(
+      const TableIdentifier& identifier, const Schema& schema, const PartitionSpec& spec,
+      const std::string& location,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  Result<std::unique_ptr<Table>> UpdateTable(
+      const TableIdentifier& identifier,
+      const std::vector<std::unique_ptr<TableRequirement>>& requirements,
+      const std::vector<std::unique_ptr<TableUpdate>>& updates) override;
+
+  Result<std::shared_ptr<Transaction>> StageCreateTable(
+      const TableIdentifier& identifier, const Schema& schema, const PartitionSpec& spec,
+      const std::string& location,
+      const std::unordered_map<std::string, std::string>& properties) override;
+
+  Result<bool> TableExists(const TableIdentifier& identifier) const override;
+
+  Status RenameTable(const TableIdentifier& from, const TableIdentifier& to) override;
+
+  Status DropTable(const TableIdentifier& identifier, bool purge) override;
+
+  Result<std::unique_ptr<Table>> LoadTable(const TableIdentifier& identifier) override;
+
+  Result<std::shared_ptr<Table>> RegisterTable(
+      const TableIdentifier& identifier,
+      const std::string& metadata_file_location) override;
+
+  std::unique_ptr<RestCatalog::TableBuilder> BuildTable(
+      const TableIdentifier& identifier, const Schema& schema) const override;
+
+ private:
+  RestCatalog(std::unique_ptr<RestCatalogConfig> config,
+              std::unique_ptr<HttpClient> client, ResourcePaths paths);
```

Comment:
Two comments:
1. client should be `std::shared_ptr<HttpClient>` because we may want to support reusing.
2. `ResourcePaths` should be created in it so remove it from ctor. ctor of ResourcePaths should not have any error.

If you are worried about any error, consider use an empty private ctor of RestCatalog and add `RestCatalog::Init` to deal with errors.